lib/datasource/deb/index.ts

```diff
@@ -0,0 +1,323 @@
+
+    let lastTimestamp: Date = null;
+    if (extractedFileExists) {
+      const stats = await fs.stat(extractedFile);
+      lastTimestamp = stats.ctime;
+    }
+
+    for (let i = 0; i < DebDatasource.compressions.length; i++) {
+      const compression = DebDatasource.compressions[i];
+      const compressedFile =
+        fullCacheDir + '/' + hashedPackageUrl + '.' + compression;
+      try {
+        const wasUpdated = await this.downloadPackageFile(
+          basePackageUrl,
+          compression,
+          compressedFile,
+          lastTimestamp
+        );
+        if (wasUpdated || !extractedFileExists) {
+          try {
+            await this.extract(compressedFile, compression, extractedFile);
+          } finally {
+            await fs.rm(compressedFile);
+          }
+        }
+        return extractedFile;
+      } catch (e) {
+        logger.warn(
+          "Couldn't download package file with compression " +
+            compression +
+            ' from ' +
+            basePackageUrl
+        );
+      }
+    }
+
+    throw 'No compression standard worked for ' + basePackageUrl;
+  }
+
+  /**
+   *
+   * @param cfg
+   * @param basePackageUrl
+   * @param compression
+   * @param lastDownloadTimestamp indicating the last time the Packages file was requested,
+   *                               set to null if no previous download has occurred
+   * @returns a boolean indicating if the file was modified
+   */
+  async downloadPackageFile(
+    basePackageUrl: string,
+    compression: string,
+    compressedFile: string,
+    lastDownloadTimestamp: Date
+  ): Promise<boolean> {
+    const packageUri = basePackageUrl + '/Packages.' + compression;
+    logger.debug(
+      'Downloading package file from ' + packageUri + ' as ' + compressedFile
+    );
+    let needsToDownload = true;
+    // we can use If-Modified-Since to avoid that we are redownloaded the file
+    // More information can be found here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since
+    if (lastDownloadTimestamp !== null) {
+      const downloadOptions: HttpOptions = {};
+      downloadOptions.headers = {
+        'If-Modified-Since': lastDownloadTimestamp.toUTCString(),
+      };
+      const headResult = await this.http.head(packageUri, downloadOptions);
+      needsToDownload = headResult.statusCode !== 304;
+    }
+
+    if (needsToDownload) {
+      const readStream = this.http.stream(packageUri);
+      const writeStream = fs.createWriteStream(compressedFile);
+
+      await fs.pipeline(readStream, writeStream);
+    }
+
+    return needsToDownload;
+  }
+
+  async probeExtractedPackage(
+    extractedFile,
+    packageName
+  ): Promise<ReleaseResult | null> {
+    // read line by line to avoid high memory consumption as the extracted Packages
+    // files can be multiple MBs in size
+    const rl = readline.createInterface({
+      input: fs.createReadStream(extractedFile),
+      terminal: false,
+    });
```

Comment:
I think we should cache the compressed file and directly stream if possible. Would sava a lot disk space. 🤔

Reply:
I agree that would indeed some disk space for the cost of computational expense. I was also thinking about removing some unused meta information from the extracted Package file which would also reduce the package file size and reduce the lookup time (which is more important I guess). What do you think?